py-polars/polars/internals/dataframe/frame.py

```diff
@@ -4368,6 +4370,16 @@ def join(
                 f"Expected 'other' join table to be a DataFrame, not a {type(other).__name__}"
             )
 
+        if args:
+            warnings.warn(
+                "`join` function signature will change in the next breaking release."
+                " After the `other` parameter, use keyword arguments to silence this warning.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+            keyword_inputs = left_on, right_on, on, how, suffix
+            left_on, right_on, on, how, suffix = [*args, *keyword_inputs[len(args) :]]  # type: ignore[misc]
```

Comment:
in case it's any use, in pandas there's this decorator

https://github.com/pandas-dev/pandas/blob/d1fc79d4a875f16b27cbd09d71b94fabcacb84b2/pandas/util/_decorators.py#L263-L337

for deprecating arguments as positional

Reply:
That looks great - I'll adapt that to our needs!

Reply:
>I definitely expect @alexander-beedie to have a better solution to this ~

I think @MarcoGorelli nailed it with this suggestion; I have nothing more to add 🤣